use JTF to avoid hang on OperationProgress API

heejaechang opened the pull request:
OperationProgress API we use to detect whether a solution is fully loaded calls into various providers that attribute to solution load. and some of them, unfortunately, require UI thread to proceed. 

according to the platform team, it is almost impossible to get rid of those restrictions since one of those providers could be a legacy component which we might not even have a control on it.

so now OperationProgress API (Workspace Status API is our abstraction) is only allowed to be used with await or JTF.Run on UI thread, we can never call it with "Wait" on UI thread otherwise, we can get into a deadlock.

this goes over all usage of OperationProgress API and uses JTF when needed.

jinujoseph commented:
@heejaechang this needs to be in preview1 ? or you are just testing ?

heejaechang commented:
@dotnet/roslyn-ide  can I get code review? @dibarbet can you take a look?

@jinujoseph for QB

dibarbet commented on src/EditorFeatures/Core/Implementation/ChangeSignature/AbstractChangeSignatureCommandHandler.cs:
```diff
@@ -2,24 +2,29 @@
+
+        protected AbstractChangeSignatureCommandHandler(IThreadingContext threadingContext)
+        {
+            _threadingContext = threadingContext;
```
I guess most of the updates to pass in the threading context to the constructors should have contracts throw if null?  I'd only push for this one in particular though to establish a pattern in the new constructor.

heejaechang replied:
@sharwell @jasonmalinowski actually is threadingcontext exist for all host? like VS for mac, unit test and etc?

jasonmalinowski replied:
IThreadingContext exists anywhere the editor does since the editor now depends on JTF, so this code here would be safe to assume it.

CyrusNajmabadi commented:
Could we get @sharwell to look at ths?  I thnk he's the strongest JTF person we have.

jasonmalinowski commented on src/EditorFeatures/CSharp/ChangeSignature/CSharpChangeSignatureCommandHandler.cs:
```diff
@@ -13,7 +14,8 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.ChangeSignature
     {
         [ImportingConstructor]
-        public CSharpChangeSignatureCommandHandler()
+        public CSharpChangeSignatureCommandHandler(IThreadingContext threadingContext) :
```
Per @CyrusNajmabadi's other PR, we're putting : before the base.

heejaechang replied:
sure

jasonmalinowski commented on src/EditorFeatures/CSharpTest/EncapsulateField/EncapsulateFieldTestState.cs:
```diff
@@ -58,7 +59,9 @@ public static EncapsulateFieldTestState Create(string markup)
-            var commandHandler = new EncapsulateFieldCommandHandler(Workspace.GetService<ITextBufferUndoManagerProvider>(),
+            var commandHandler = new EncapsulateFieldCommandHandler(
+                Workspace.ExportProvider.GetExportedValue<IThreadingContext>(),
+                Workspace.GetService<ITextBufferUndoManagerProvider>(),
```
Is there a difference between GetService and ExportProvider.GetExportedValue?

sharwell replied:
ExportProvider.GetExportedValue obtains the single value directly from the MEF catalog. GetService gets a workspace service in the context of a specific workspace (accounting for things like ServiceLayer).

heejaechang replied:
I got confused as well. so this GetService actually return MEF Service like @jasonmalinowski indicates not Workspace services...

Workspace.Services.GetService returning workspace service and Workspace.GetService returning MEF service is so confusing..

heejaechang replied:
there is so many references on GetService so i won't change its naming in this PR...

sharwell replied:
> so this GetService

Ah, yes, this is the `GetService` defined in test code.

jasonmalinowski replied:
@heejaechang Sounds good, do we want a follow-up then to just always require Workspace.ExportProvider.GetExportedValue? Although longer I'll agree it's less confusing!

dibarbet commented on src/EditorFeatures/CSharp/ChangeSignature/CSharpChangeSignatureCommandHandler.cs:
```diff
@@ -13,7 +14,8 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.ChangeSignature
     {
         [ImportingConstructor]
-        public CSharpChangeSignatureCommandHandler()
+        public CSharpChangeSignatureCommandHandler(IThreadingContext threadingContext) :
```
minor - I believe @CyrusNajmabadi just changed a bunch of these to move the colon to the line with base.
Until there's an actual style setting for it though I don't mind.

jasonmalinowski commented on src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:
```diff
@@ -10371,8 +10372,8 @@ public void ExtractMethodCommandDisabledInSubmission()
                 var textView = workspace.Documents.Single().GetTextView();
 
                 var handler = new ExtractMethodCommandHandler(
+                    exportProvider.GetExportedValue<IThreadingContext>(),
```
Why not GetService like the rest?

sharwell replied:
➡️ `IThreadingContext` is not a workspace service

jasonmalinowski commented on src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:
```diff
@@ -169,7 +168,9 @@ public bool TryGetTelemetryId(out Guid telemetryId)
                     //       text on the wait dialog window
-                    _workspaceStatusService.WaitUntilFullyLoadedAsync(cancellationToken).Wait(cancellationToken);
+                    // 
+                    //       this also needs to run under threading context otherwise, we can deadlock on VS
```
Any idea how we can ensure people *do* use JTF with this one as opposed to assuming Roslyn style waiting?

heejaechang replied:
I am not sure. I think at some point, we might always need to use _threadingContext for wait and based on host, it might or might not use JTF for waiting...

jasonmalinowski commented on src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs:
```diff
@@ -89,5 +90,16 @@ private static IEnumerable<AbstractFormattingRule> GetFormattingRules(Document d
+        /// Get <see cref="Document"/> from <see cref="Text.Extensions.GetOpenDocumentInCurrentContextWithChanges(ITextSnapshot)"/>
+        /// once <see cref="IWorkspaceStatusService.WaitUntilFullyLoadedAsync(CancellationToken)"/> returns
+        /// </summary>
+        public static Document GetFullyLoadedOpenDocumentInCurrentContextWithChanges(
```
I like that you have this method that internally hides the async if only to prevent people from trying to do a .Wait() without realizing it. What do we want to do with the other one that's the async method? Does it minimally need a comment that you need a JTF wait for it?

sharwell replied:
This is an anti-pattern. The sync-to-async transition is undesirable in the implementation and should not be forced on callers, especially callers that are already asynchronous. The transition should always be moved as far to the callers as possible, and eventually eliminated. The risk of `Wait` should not exist because `Wait` is not allowed in Visual Studio. An exception is claimed for provable cases that do not require the UI thread, but this is not validated or enforced so it repeatedly leads to bugs. At a sync-to-async boundary, `Wait` should always be assumed to be a deadlock risk.

jasonmalinowski commented on src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs:
```diff
@@ -89,5 +90,16 @@ private static IEnumerable<AbstractFormattingRule> GetFormattingRules(Document d
+        public static Document GetFullyLoadedOpenDocumentInCurrentContextWithChanges(
+            this ITextSnapshot snapshot, IUIThreadOperationContext operationContext, IThreadingContext threadingContext)
+        {
+            return threadingContext.JoinableTaskFactory.Run(() =>
```
Just like we have WaitAndGetResult() assert if you're on a foreground thread to make sure nobody is accidentally blocking up backgrounds threads when they could have used the async form, does this need the same?

sharwell replied:
:memo: This comment will likely be obsolete once https://github.com/dotnet/roslyn/pull/36864#discussion_r299219474 is addressed.

heejaechang replied:
I wrote big comment on why I like my approach. but github seems not showing the comments...

basically, I think the way it is done is safer than forcing everything to call JTF.Run(...)

jasonmalinowski commented on src/EditorFeatures/Test2/Rename/RenameCommandHandlerTests.vb:
```diff
@@ -18,8 +19,9 @@ Namespace Microsoft.CodeAnalysis.Editor.UnitTests.Rename
-            Return New RenameCommandHandler(workspace.GetService(Of InlineRenameService),
-                                            workspace.GetService(Of IEditorOperationsFactoryService))
+            Return New RenameCommandHandler(
+                workspace.ExportProvider.GetExportedValue(Of IThreadingContext)(),
```
Why mixing ExportProvider and GetService?

sharwell replied:
➡️ `IThreadingContext` is exported directly (not a workspace service)

jasonmalinowski commented on src/EditorFeatures/Test2/Rename/RenameTagProducerTests.vb:
```diff
@@ -565,8 +566,9 @@ class C
-                Dim commandHandler As New RenameCommandHandler(workspace.GetService(Of InlineRenameService),
-                                                               workspace.GetService(Of IEditorOperationsFactoryService))
+                Dim commandHandler As New RenameCommandHandler(
+                    workspace.ExportProvider.GetExportedValue(Of IThreadingContext)(),
```
Why mixing ExportProvider and GetService?

sharwell replied:
`IThreadingContext` is exported directly (not a workspace service)

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:
```diff
@@ -76,6 +76,9 @@ public Service(IAsyncServiceProvider2 serviceProvider, IAsynchronousOperationLis
 
+            // unfortunately, IVsOperationProgressStatusService requires UI thread to let project system to proceed to next stages.
+            // this method should only be used with either await or JTF.Run, it should be never used with Task.Wait otherwise, it can
+            // deadlock
```
Move this to a doc comment so there's some chance that somebody will actually see this at a call site.

jasonmalinowski replied:
The comment should probably also be moved to the workspace service itself (or some comment like that) since most consumers are going to see it there -- nobody really sees the direct type like this.

sharwell replied:
📝 All asynchronous APIs defined for Visual Studio fall into this category unless explicitly stated otherwise.

heejaechang replied:
it is not all async API from VS. synchronous API from VS can also fall into this category. so any call to VS API from async caller that we do "Wait" on UI thread can cause an issue.

and I am not solving that root issue in this PR. if we want to solve that issue, feel free to open an issue where we can discuss how we can embrass JTF more deepfully in Roslyn.

jasonmalinowski commented:
And oh: it's otherwise definitely good ignoring my doc comment concern. Fully happy to approve when fixed, and if this needs to go through shiproom people are welcome to override my block if necessary, assuming it's been addressed.

sharwell commented on src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs:
```diff
@@ -89,5 +90,16 @@ private static IEnumerable<AbstractFormattingRule> GetFormattingRules(Document d
+        /// Get <see cref="Document"/> from <see cref="Text.Extensions.GetOpenDocumentInCurrentContextWithChanges(ITextSnapshot)"/>
+        /// once <see cref="IWorkspaceStatusService.WaitUntilFullyLoadedAsync(CancellationToken)"/> returns
+        /// </summary>
+        public static Document GetFullyLoadedOpenDocumentInCurrentContextWithChanges(
```
❗️ This method had the correct signature _before_ this change. Sync/async transitions should always be pushed to the callers, not to the callees.

heejaechang replied:
"The transition should always be moved as far to the callers as possible" is true in normal Task case since Wait on a task could lead to deadlock, VS or not. especially due to UI thread in VS since UI thread act as a single lock.

I am not sure whether that is true for JTF since one of the purposes of JTF is to hide such detail I believe. in practice, that might be the whole purpose of people why they are using it.

otherwise, none of VS APIs, I mean literally none of APIs we use that internally use JTF should not use it. since that basically violates what you said above. we currently have no idea which VS API we uses internally use JTF (especially JTF.Run) and asked just use JTF always to make whole thing works in VS.

if we are in the world you stated above, we should know exactly what VS APIs are using JTF inside and we are required to use them under JTF in UI thread.

but we don't live in such world. and Roslyn which doesn't adapt JTF fully make things more complicated. 

we already have many APIs which support sync and async version of APIs such as Read and ReadAsync or GetRoot and GetRootAsync and more.

I want people to consider this API as such. sure, people might use sync version in async method and then call Wait on the async method causing us a deadlock. but conceptually same problem already exist in other cases such as sync method calling async version of GetRootAsync which will cause thread starvation. and we provide sync API so that caller has an option not to call async version in sync code path.

I would like users to use it that way. they don't need to know about the implementation detail and as long as they use sync version in sync code path, they are fine. sure there is always corner case where sync code path is nested deep in sync code path that got called with async method with async method at the top with Wait.

to help avoiding such case, it is an extension method and actual service doesn't have sync API. also, I dont think the concern above can be removed by forcing people to repeatedly write the same JTF.Run code. now they have the risk of not calling this in JTF.Run and cause hang.

on whether we should always use JTF.Run and never wait. that is not how Roslyn is written. and until we get rid of all Wait from Roslyn, Wait/Result is valid option for Roslyn.

heejaechang commented on src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:
```diff
@@ -169,7 +168,9 @@ public bool TryGetTelemetryId(out Guid telemetryId)
+                    // 
+                    //       this also needs to run under threading context otherwise, we can deadlock on VS
+                    ThreadingContext.JoinableTaskFactory.Run(() => _workspaceStatusService.WaitUntilFullyLoadedAsync(cancellationToken));
                 }
```
by the way, why we leak that we are using JTF in threadingcontext? it could just expose ThreadingContext.Wait so we just use it for every case we use Wait or Result?

heejaechang commented:
@jasonmalinowski I added main thread check and updated doc comment.

@sharwell I wrote why I don't want to spread JTF all over the places. for me, what you meant is "let's use JTF for every places we use Wait", and if you want to that, open an issue and bring that to design meeting.

jasonmalinowski commented on src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs:
```diff
@@ -91,7 +91,18 @@ internal async Task SearchAsync()
-                    _callback.Done();
+                    var service = _solution.Workspace.Services.GetService<IWorkspaceStatusService>();
+                    if (!await service.IsFullyLoadedAsync(_cancellationToken).ConfigureAwait(false) &&
+                        _callback is INavigateToCallback2 callback2)
```
Should this do the IsNaviageToCallback2 check first since that's the cheaper one?

heejaechang replied:
sure.

heejaechang commented:
closed and reopen to re-run tests.

heejaechang commented:
I logged this issue - https://github.com/dotnet/roslyn/issues/36937 - so we can decide on JTF usage in Roslyn. so that we dont have multiple different or opposite opinions on where it can be used.
